language/move-prover/tests/sources/functional/global_vars.move

```diff
@@ -1,162 +1,145 @@
+    use Std::Signer;
 
-    spec module {
-        pragma verify = true;
-    }
+    // ================================================================================
+    // Counting
 
     spec module {
-        global sum_of_T: u64;
+        global sum_of_T: u64 = 0;
     }
 
-    resource struct T {
+    struct T has key {
       i: u64,
     }
-    spec T {
-      invariant pack sum_of_T = sum_of_T + i;
-      invariant unpack sum_of_T = sum_of_T - i;
+
+    fun add() acquires T {
+        borrow_global_mut<T>(@0).i = borrow_global_mut<T>(@0).i + 1
+    }
+    spec add {
+        update sum_of_T = sum_of_T + 1;
     }
 
-    // ----------
-    // Pack tests
-    // ----------
+    fun sub() acquires T {
+        borrow_global_mut<T>(@0).i = borrow_global_mut<T>(@0).i - 1
+    }
+    spec sub {
+        update sum_of_T = sum_of_T - 1;
+    }
 
-    fun pack_correct(): T {
-        T {i: 2}
+    fun call_add_sub() acquires T {
+        add(); add(); sub();
     }
-    spec pack_correct {
-        ensures sum_of_T == old(sum_of_T) + 2;
+    spec call_add_sub {
+        ensures sum_of_T == 1;
     }
 
-    fun pack_incorrect(): T {
-        T {i: 2}
+    fun call_add_sub_invalid() acquires T {
+        add(); sub(); add();
     }
-    spec pack_incorrect {
-        ensures sum_of_T == old(sum_of_T) + 1;
-        ensures result.i == 2;
+    spec call_add_sub_invalid {
+        ensures sum_of_T == 2;
     }
 
+    // ================================================================================
+    // Access Control
 
-    // ------------
-    // Unpack tests
-    // ------------
+    spec module {
+        // Indicates whether a specific access has been verified. This is kept intentionally
+        // uninitialized so the prover will find situations where this is false but access is required.
+        global access_verified: bool;
+    }
 
-    fun unpack_correct(t: T): u64 {
-        let T {i: x} = t;
-        x
+    fun assert_access(s: &signer) {
+        // Do some assertions which validate access
+        assert(Signer::address_of(s) == @0, 1);
     }
-    spec unpack_correct {
-        ensures sum_of_T == old(sum_of_T) - old(t.i);
-        ensures result == old(t.i);
+    spec assert_access {
+        aborts_if Signer::spec_address_of(s) != @0;
+        update access_verified = true;
     }
 
-    fun unpack_incorrect(t: T): u64 {
-        let T {i: x} = t;
-        x
+    fun requires_access() {
+        // Do some things which require access to be validated.
     }
-    spec unpack_incorrect {
-        ensures sum_of_T == old(sum_of_T);
-        ensures result == old(t.i);
+    spec requires_access {
+        requires access_verified;
     }
 
-    // ------------
-    // Update tests
-    // ------------
+    fun do_privileged(s: &signer) {
+        assert_access(s);
+        requires_access();
+    }
 
-    fun update_valid_still_mutating(t: &mut T) {
-        t.i = t.i + 3;
+    fun do_privileged_invalid(_s: &signer) {
```

Comment:
Since the value of the ghost variable `access_verified` persists, I just see a similar situation to the example we discussed on Slack:
`do_privileged_invalid` fails by itself, but is ok in:
```
fun h(s: &signer) {
  do_privileged(s);
  do_privileged_invalid(s);
}
```
I think the Prover's behavior here is intended and has no issue.

Reply:
The difference is that you can observe this behavior now in the same model. Within the same model, one can see that do_privleged_invalid behaves one time this way, and the other time a different way, because it now depends on *state*. The issue we discussed on Slack is that because is_signer has no state dependencies, within the same model it always behaves the same, and you cannot observe the expected thing in a unified model.